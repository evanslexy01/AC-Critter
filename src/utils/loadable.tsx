import * as React from 'react';

export const lazyLoad = <T extends Promise<any>, U extends React.ComponentType<any>>(
  importFunc: () => T,
  selectorFunc: (s: T extends Promise<infer P> ? P : never) => U,
  fallback: React.ReactNode = null
) => {
  const LazyComponent = React.lazy(() =>
    importFunc().then(module => ({ default: selectorFunc(module) }))
  );
  return (props: React.ComponentProps<U>): JSX.Element => (
    <React.Suspense fallback={fallback}>
      <LazyComponent {...props} />
    </React.Suspense>
  );
};
