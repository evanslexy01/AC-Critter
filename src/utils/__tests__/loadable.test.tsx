import { lazyLoad } from '../loadable';
import { render } from '@testing-library/react';
import * as React from 'react';

const LazyComponenWithDefaultExport = lazyLoad(
  () => import('../../pages/CollectionsPage'),
  m => m.default
);

const LazyComponentWithExportedFunction = lazyLoad(
  () => import('../../pages/CollectionsPage'),
  m => m.default
);

const LazyComponentWithFallback = lazyLoad(
  () => import('../../pages/CollectionsPage'),
  m => m.default,
  <div>Loading</div>
);

describe('loadable', () => {
  it('should render null by default', async () => {
    const {
      container: { firstChild },
    } = render(<LazyComponenWithDefaultExport />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render null by default with empty options', async () => {
    const {
      container: { firstChild },
    } = render(<LazyComponentWithExportedFunction />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render fallback if given one', async () => {
    const {
      container: { firstChild },
    } = render(<LazyComponentWithFallback />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render LazyComponent after waiting for it to load', async () => {
    const {
      container: { firstChild },
    } = render(<LazyComponentWithExportedFunction />);
    LazyComponentWithExportedFunction({});
    expect(firstChild).toMatchSnapshot();
  });
});
