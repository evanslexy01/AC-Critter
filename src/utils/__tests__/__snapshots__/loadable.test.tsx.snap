// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadable should render LazyComponent after waiting for it to load 1`] = `
<div
  class="page page-hide-left"
>
  <nav
    class="ac-header"
  >
    <a>
      <img
        alt="settings"
        src="[object Object]"
      />
    </a>
    <h1>
      Collections
    </h1>
    <a>
      <img
        alt="critters"
        src="[object Object]"
      />
    </a>
  </nav>
  <div
    class="content"
  >
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Bugs: 
        110
        /
        114
      </h3>
    </div>
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Sea Life: 
        110
        /
        114
      </h3>
    </div>
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Fish: 
        110
        /
        114
      </h3>
    </div>
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Fossils: 
        110
        /
        114
      </h3>
    </div>
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Songs: 
        110
        /
        114
      </h3>
    </div>
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Art: 
        110
        /
        114
      </h3>
    </div>
    <div
      class="bubble"
    >
      <div
        class="icon-holder"
      >
        <img
          alt="Squid"
          class="icon"
          src="[object Object]"
        />
      </div>
      <h3
        class="title"
      >
        Gyroids: 
        110
        /
        114
      </h3>
    </div>
  </div>
</div>
`;

exports[`loadable should render fallback if given one 1`] = `
<div>
  Loading
</div>
`;

exports[`loadable should render null by default 1`] = `null`;

exports[`loadable should render null by default with empty options 1`] = `null`;
